### PR TITLE
Add `project_from_id` provider-defined function

### PR DESCRIPTION
**Pull request by SarahFrench**
To be merged after https://github.com/GoogleCloudPlatform/magic-modules/pull/10013 Edit: done!

This PR:
- adds the new project_from_id function to the providers
- adds documentation for the Registry
- adds unit tests to test reusable logic that will be used by other X_from_id functions
- adds unit tests for function calls to the project_from_id function specifically - test good vs bad input
- adds acceptance tests that demonstrate the function is usable with the provider and can be used in a single apply alongside provisioning a resource

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
provider: added provider-defined function `project_from_id` for retrieving the project id from a resource's self link or id
```

### CODE REVIEW

File: mmv1/third_party/terraform/functions/project_from_id_internal_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/functions/project_from_id_internal_test.go:1]: I've introduced these tests following the advice in https://developer.hashicorp.com/terraform/plugin/framework/functions/testing#unit-testing

**Comment by SarahFrench** [comment]
This PR needs to be rebased to include https://github.com/GoogleCloudPlatform/magic-modules/pull/10013

**Comment by SarahFrench** [comment]
Rebased following the 2 blocking PRs being merged into the feature branch

**Comment by SarahFrench** [comment]
@rileykarson The "Error: Missing newline after argument" error in the `TestAccProviderFunction_project_from_id` acceptance test failure is what's seen when people try to use a provider-defined function with older versions of Terraform.

Checking out the [v1.8.0-alpha20240216](https://github.com/hashicorp/terraform/releases/tag/v1.8.0-alpha20240216) tag and building Terraform locally will enable manually running this test

File: mmv1/third_party/terraform/website/docs/functions/project_from_id.html.markdown [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+## Signature
+
+```text
+project_from_id(id string) string
+```
+
+## Arguments
+
+1. `id` (String) A string of a resource's id, a resource's self link, or an OP style resource name. For example, these are all valid values:
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/website/docs/functions/project_from_id.html.markdown:38]: I don't know what the correct term to use here is w.r.t "OP style resource name". I got that phrase from here (https://github.com/hashicorp/terraform-provider-google/issues/16244#issuecomment-1867033535) and I've found [docs for Cloud Asset Inventory that list this type of resource name](https://cloud.google.com/asset-inventory/docs/resource-name-format) but I'm not really satisfied. I'm not sure what OP stands for above!

Comment (SarahFrench) [reply mmv1/third_party/terraform/website/docs/functions/project_from_id.html.markdown:38]: https://en.wikipedia.org/wiki/URL#prurl is closest I've got

File: mmv1/third_party/terraform/website/docs/functions/project_from_id.html.markdown [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      source = "hashicorp/google"
+    }
+  }
+}
+
+# Value is "my-project"
+output "function_output" {
+  value = provider::google::project_from_id("https://www.googleapis.com/compute/v1/projects/my-project/zones/us-central1-c/instances/my-instance")
+}
+```
```
[the remark this answers is not in the document]
Comment (rileykarson) [reply mmv1/third_party/terraform/website/docs/functions/project_from_id.html.markdown:28]: It seems reasonable to document for cases where folks are using just TPGB in a configuration!

File: mmv1/third_party/terraform/website/docs/functions/project_from_id.html.markdown (hunk 1 above)
Comment (SarahFrench) [reply mmv1/third_party/terraform/website/docs/functions/project_from_id.html.markdown]: Thank you for this - I've updated the documentation using that new phrasing!

File: mmv1/third_party/terraform/website/docs/functions/project_from_id.html.markdown (hunk 2 above)
[the remark this answers is not in the document]
Comment (SarahFrench) [reply mmv1/third_party/terraform/website/docs/functions/project_from_id.html.markdown:30]: I've updated the docs in [431dd77](https://github.com/GoogleCloudPlatform/magic-modules/pull/10021/commits/431dd77ab017e25354dd7309ef792858ff2a6d78) to have an example block per provider.

The terraform block needs to be present to use provider-defined functions, so I think it's worth including the full snippet twice, despite the fact they only differ by adding `-beta`:


Screenshot 2024-02-26 at 16 53 08

**Comment by SarahFrench** [comment]
@rileykarson could you please review this PR? I'd like to get this PR (plus other PRs for functions to get X from id) into the feature branch so I can then update the feature branch to navigate [a breaking change recently introduced into the plugin-framework in v1.6.0](https://github.com/hashicorp/terraform-plugin-framework/releases/tag/v1.6.0) that affects provider-defined functions.

**Review by rileykarson — APPROVED** [review]

**Pull request merged**